detect/detect.go

```diff
@@ -67,3 +67,58 @@ func ExtensionDataFormat(path string) (format dataset.DataFormat, err error) {
 		return dataset.UnknownDataFormat, fmt.Errorf("unsupported file type: '%s'", ext)
 	}
 }
+
+// ErrInvalidTabularData indicates non-tabular data in a context that expects
+// tabular input
+var ErrInvalidTabularData = errors.New("invalid tabular data")
+
+// TabularSchemaFromTabularData infers a basic tabular JSON schema from go types
+// it only works in the narrow case where the source data is known to be tabular
+// but lacks a schema to describe it
+// given the lack of metadata, these schema should be used primarily for
+// machine purposes
+func TabularSchemaFromTabularData(source interface{}) (map[string]interface{}, error) {
+	schema := map[string]interface{}{}
+	switch data := source.(type) {
+	case []interface{}:
+		schema["type"] = "array"
+		items := map[string]interface{}{}
+		if len(data) == 0 {
+			return nil, fmt.Errorf("%w: missing row data", ErrInvalidTabularData)
+		}
+
+		switch ent := data[0].(type) {
+		case []interface{}:
+			items["type"] = "array"
+			cols := make([]interface{}, len(ent))
+			for i, v := range ent {
+				cols[i] = map[string]interface{}{
+					"title": fmt.Sprintf("col_%d", i),
+					"type":  goDataType(v),
+				}
+			}
+			items["items"] = cols
+		default:
+			return nil, fmt.Errorf("%w: array schemas must use an inner array for rows", ErrInvalidTabularData)
+		}
+		schema["items"] = items
+	case map[string]interface{}:
+		return nil, fmt.Errorf("%w: cannot interpret object-based tabular schemas", ErrInvalidTabularData)
+	}
+
+	return schema, nil
+}
+
+func goDataType(v interface{}) string {
+	switch v.(type) {
+	case int64, int, uint, float64:
```

Comment:
Not sure if we should also include float32, should probably also include other variations of int, but I'm most worried about float32 as we have it pop up in other places as well.

Reply:
Makes perfect sense. I’ll fix!